message view: Scroll on document instead of ".app".

ryanreh99 opened the pull request:
Fixes #9911.
The compose box also had a partial overlap with the scrollbar. This fixes it.
![header](https://user-images.githubusercontent.com/40122794/94192125-240f7e00-fecc-11ea-865a-c04de8a0440e.png)

- [ ] It looks like we used to use the browser scrollbar ([`dedd3c`](dedd3cf03cdc7291817da9ba370e806d2e069267)) but then `#panels` wasn't implemented,
used `sticky` as I couldn't find simpler way without including javascript changes and
we could polyfill it as we are doing so for the `frb`.

WIP as I'm looking into this and haven't tested on IE11+.

**Testing Plan:** <!-- How have you tested? -->
For scroll changes, grepped for `.offset().top` and `$(".app")` then manually tested the events which called those functions.
For UI changes, added extra users, streams and messages using `populate_db`.

andersk commented on static/styles/zulip.css:
```diff
@@ -33,6 +29,8 @@ html {
 }
 
 body {
+    overflow-x: hidden;
+    overflow-y: auto;
```
For standards-mode pages like ours, the scrollbar [belongs on `html`](https://developer.mozilla.org/en-US/docs/Web/API/document/scrollingElement), not `body`.

andersk commented on static/styles/zulip.css:
```diff
@@ -295,25 +296,14 @@ p.n-margin {
-    height: 100%;
-    overflow-y: scroll;
     z-index: 99;
     -webkit-overflow-scrolling: touch;
```
We can remove this `-webkit-overflow-scrolling` (it should already be the default on the root scrollbar).

andersk commented on static/js/message_viewport.js:
```diff
@@ -41,14 +44,14 @@ exports.message_viewport_info = function () {
     // and we err on the side of saying that we are at
     // the bottom.
-    return bottom + 2 >= full_height;
+    return bottom + 5 >= full_height;
```
Hmm, what happened here?

ryanreh99 replied:
I think this a bug unrelated to this PR where the current logic does not take into account the `#panels` element.
Thus the `message_viewport.at_bottom` always returns `false` when the panel is displayed.
The bugs due to this are also reproducible on CZO.

For example, using the `PageDown` key, the condition on [this line](https://github.com/zulip/zulip/blob/master/static/js/navigate.js#L113) does not satisfy.
Thus in the gif, the 1st key press causes the scrollbar to shift down and the 2nd key press makes the pointer select the bottom most message which should have been done by the first key press.
On removing the panel, the behaviour gets fixed.
![PgDn](https://user-images.githubusercontent.com/40122794/95354429-0be32a00-08e2-11eb-962c-cebb13f3a847.gif)

There is also a gap below the sidebars if the page is loaded with the panel and then removed.
The changes in this commit fixes it.

I'll revert this change and test this commit without the panel so that behaviour remains the same.
And create another commit for the changes when panel is displayed.

andersk commented:
This seems to work pretty well. I noticed two issues, both related to modals:

* If you open a modal (e.g. settings) and then close it, the view unexpectedly scrolls up to the top.
* The scrollbar is no longer disabled while a modal is active. We’ll need to disable it to protect the user from accidentally marking messages as read. This may be a bit tricky to do without visibly shifting the page content around ([Bootstrap does it](https://github.com/twbs/bootstrap/blob/v4.5.3/js/src/modal.js#L478-L545) by hiding the scrollbar and adding `padding-right` to various elements).

ryanreh99 commented:
Rebased the branch, mostly just resolved merge conflicts.
And fixed the 2 issues in the review comment above.

`scroll_bar.js` contains the to logic to enable / disable scrolling.

Tried other CSS related approaches to disable scrolling, however all of them cause the view to scroll at the top.
This causes more messages to be fetched affecting the `last_scroll_position` calculation.
Thus 1 behaviour change has been made - We do not send a fetch messages request when an overlay is open.

Hi @amanagr could you also have a look.


https://user-images.githubusercontent.com/40122794/149632895-3161189b-ba60-496d-b1c3-8bd0c4329f1e.mov

amanagr commented:
After understanding the change that need to be done, I figured that these changes are  a bit incomplete.

What we changed:
`overflow-y: scroll` on `html` instead of `app`.

Impact:
`.floating_recipient`, navbar and `.selectable_row` have their `.offset()` properties (top, bottom) increased by `$(window).scrollTop()`.

`#compose` offset didn't change.

So, we need to find places where `offset` of these elements are being used, not necessarily using the `offset` method and change it too.

amanagr commented on static/js/floating_recipient_bar.js:
```diff
@@ -10,9 +10,7 @@ let is_floating_recipient_bar_showing = false;
-        elem.offset().top -
-        $("#message_view_header").safeOuterHeight() -
-        $("#navbar_alerts_wrapper").height()
+        elem.offset().top - $("#navbar-middle").offset().top - $("#navbar-middle").safeOuterHeight()
```
Why use `$("#navbar-middle").offset().top`? We should use `$(window).scrollTop()` to simplify understanding of the change.

ryanreh99 replied:
This was done based on this review comment : https://github.com/zulip/zulip/pull/16411#discussion_r494586964

amanagr commented on static/js/overlays.js:
```diff
@@ -77,6 +82,7 @@ export function active_modal() {
 }
 
 export function open_overlay(opts) {
+    scroll_bar.disable_scrolling();
     popovers.hide_all();
```
Some overlays use `micromodal` library, we should disable scrolling in them too.

amanagr commented on static/styles/zulip.css:
```diff
@@ -25,23 +25,28 @@ $right_sidebar_width: 250px;
 
+html {
+    overflow-x: hidden;
+    overflow-y: auto;
```
We should use `overflow-y: scroll` otherwise it has weird scrollbar appear - disappear effect.

ryanreh99 replied:
Thanks! Tested this, there was a bug in narrow widths where the content shifted a bit on opening modals due to this scrollbar effect.
This fixes it.

timabbott commented:
We merged #20699, which had pieces of this plan.  I think everyone's in agreement that if we can get this fully working, it'd be even better, so @ryanreh99 I encourage you to keep working on this as you have time :)

amanagr commented:
@ryanreh99 can I take over this PR? Let me know if have done any work after my review and would like to update this PR. :)

ryanreh99 commented:
Hey @amanagr  , sorry about the delay.
Sure please go ahead.
I can test the future updates.

> So, we need to find places where offset of these elements are being used, not necessarily using the offset method and change it too. You missed a few places like maybe_scroll_up_selected_message in compose_actions and start_edit_maintaining_scroll in message_edit, etc.



Yeah, I had only grepped and checked all the places where `offset` and `$(".app")` were used in calculation. 
The `maybe_scroll_up_selected_message` and `start_edit_maintaining_scroll` places didn't have to be updated because both the differences cancelled out in subtraction.


I wonder if we should just modify the `.offset().top` to always subtract the `scrollTop()` from the result
( like how we do for `safeOuterHeight` ).
This would help in reduce the no. of lines changed and if we missed any place.

timabbott commented:
When we get this working, we should re-test https://chat.zulip.org/#narrow/stream/101-design/topic/Images.20in.20a.20line.20vs.20row/near/1326556.

amanagr commented:
Not ready for review, only rebased on master.

amanagr commented on frontend_tests/node_tests/hotkey.js:
```diff
@@ -62,6 +62,7 @@ const overlays = mock_esm("../../static/js/overlays", {
     drafts_open: () => false,
```
Same commit in #21672

timabbott replied:
Cool, this is merged so we can drop this on rebase.

amanagr commented:
I tried to set $message_pane as "html" but it resulted in app having so many bugs that it was hard to identify how things got regressed. This PR is actually is in a good state where I don't see any issues. We should aim to get this test deployed.

amanagr commented:
> When we get this working, we should re-test https://chat.zulip.org/#narrow/stream/101-design/topic/Images.20in.20a.20line.20vs.20row/near/1326556.

I don't see any difference.

andersk commented:
Some more little issues:

* If you set your **Default view** to **All messages**, and leave **Escape key navigates to default view** enabled, then pressing <kbd>Esc</kbd> navigates to `#` and scrolls up to the top.
* If you’re in a message view not scrolled to the bottom, opening or closing a modal re-triggers the **Scroll to bottom** button to fade in.

amanagr commented:
I fixed the first issue by enabling scrolling after we close the modal. I cannot reproduce the second issue. Maybe there is a better reproducer?

amanagr commented on static/third/bootstrap-tooltip/tooltip.js:
```diff
@@ -167,8 +167,8 @@
                  to also use the "no_arrow_popover" template.
               */
-              if (top < 0) {
-                  top = 0;
+              if (top < $(window).scrollTop()) {
+                  top = $(window).scrollTop();
```
Added some new changes to fix placement of popovers.

andersk commented:
Both issues still reproduce.

> * If you set your **Default view** to **All messages**, and leave **Escape key navigates to default view** enabled, then pressing <kbd>Esc</kbd> navigates to `#` and scrolls up to the top.

Clarification: this happens when you’re already in the all messages view. Also, in Chrome, you may need to explicitly click on **All messages** before reproducing.

> * If you’re in a message view not scrolled to the bottom, opening or closing a modal re-triggers the **Scroll to bottom** button to fade in.

Correction: in a message view scrolled to somewhere in the middle (neither the top nor the bottom).

amanagr commented:
Noticed something more weird while trying to reproduce above, where the app just froze and scrolling just didn't work. Not sure what caused it.

amanagr commented:
Closing in favour of  #25697
